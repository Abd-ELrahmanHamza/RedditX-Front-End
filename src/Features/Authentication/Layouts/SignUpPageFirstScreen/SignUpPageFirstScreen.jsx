/* eslint-disable no-unused-vars */
/* eslint-disable react-hooks/exhaustive-deps */
import { useEffect } from "react";


import { FaFacebookSquare } from "react-icons/fa";

import FacebookLogin from "react-facebook-login";

import { FcGoogle } from "react-icons/fc";

import { signInWithGooglePopup } from "Features/Authentication/Utils/Firebase";

import RandomUserName from "Features/Authentication/Utils/RandomUserName";

import FormInputPageCom from "Features/Authentication/Components/FormInputPageCom/FormInputPageCom";
import Button from "../../Components/Button/Button";

import {
  LogInContainer,
  PageContainer,
  Photo,
  AuthContainer,
  AuthContainerDiv,
  All,
  ButtonsContainer,
  Privacy,
  ErrorParagraph,
  SignInWithGoogle,
  SignInWithFacebook,
  OrHeader,
  AuthParagraph,
  AuthHeader,
  Forget,
} from "./SignUpPageFirstScreen.styled";

const USER_EMAIL =
  /[a-zA-Z0-9._-]{3,}@[a-zA-Z0-9._-]{3,}[.]{1}[a-zA-Z0-9._-]{2,}/;

/**
 * SignUpPageFirstScreen component that is used in Signup page
 * @param {boolean} validEmail Prop to know if the email is valid or not
 * @param {Function} setValidEmail Function to set the validity of the email
 * @param {boolean} secondScreen Prop to know what screen should be shown (the choose email screen or choose userName and password screen)
 * @param {Function} setSecondScreen Function to set the state of secondScreen
 * @param {boolean} initialFocus Prop to know if the user made at least one focus on the input field or not
 * @param {Function} setInitialFocus Function to set the state of initialFocus
 * @param {Function} changeUserName Function to change the random userName
 * @param {Object} formFields Object contain the values of input fields
 * @param {Function} setFormFields Function to update the values of input fields
 * @param {String} errMsg The error message that appear under the input field
 * @param {Function} setErrMsg Function to set the error message
 * @param {Function} setValidUserName Function to set the validity of the userName
 * @param {Function} setSug1 Function to set the random userName number 1
 * @param {Function} setSug2 Function to set the random userName number 2
 * @param {Function} setSug3 Function to set the random userName number 3
 * @param {Function} setSug4 Function to set the random userName number 4
 * @param {Function} setSug5 Function to set the random userName number 5
 * @returns {React.Component}  SignUpPageFirstScreen component that is used in Signup page
 */

const SignUpPageFirstScreen = ({
  formFields,
  setFormFields,
  validEmail,
  setValidEmail,
  secondScreen,
  setSecondScreen,
  initialFocus,
  setInitialFocus,
  setValidUserName,
  errMsg,
  setErrMsg,
  setSug1,
  setSug2,
  setSug3,
  setSug4,
  setSug5,
}) => {
  const { email } = formFields;

  /**
   * useEffect for email field to check if the email that the user entered is valid or not
   */
  useEffect(() => {
    if (email.length > 0) {
      setInitialFocus(false);
      setErrMsg("Not a valid email address");
    } else if (email.length === 0) {
      setErrMsg("Please enter an email address to continue");
    }

    setValidEmail(USER_EMAIL.test(email));

    setSug1(RandomUserName());
    setSug2(RandomUserName());
    setSug3(RandomUserName());
    setSug4(RandomUserName());
    setSug5(RandomUserName());
  }, [email]);

  /**
   * Function to handle the submit of the form of signup
   * @param {*} event
   */
  const handleSubmit = async (event) => {
    event.preventDefault();
  };

  /**
   * Function to handle any change on the input field of the signup form (check if the userName or the email or the password is valid or not)
   * @param {*} event
   */
  const handleChange = (event) => {
    const { name, value } = event.target;
    setFormFields({ ...formFields, [name]: value });
    if (name === "email") {
      setValidEmail(USER_EMAIL.test(email));
    }
  };

  /**
   * Function to handle the response coming from sign in with google
   * @param {*} response
   */
  const handleCallbackResponse = (response) => {};

  /**
   * Function to check email by sending request to the api and make sure that the email is good to show the second screen after that
   * @param {string} mail The email that the user entered
   */
  const checkEmail = (mail) => {
    /* After mail checked */

    setInitialFocus(true);
    setSecondScreen(true);
  };

  /**
   * Adding some configurations to the signIn with google feature
   */
  useEffect(() => {
    /* global google */

    google.accounts.id.initialize({
      client_id:
        "598360538255-siv9rljce260ek4mvdsu2fk63h3u1g3b.apps.googleusercontent.com",
      callback: handleCallbackResponse,
    });

    google.accounts.id.renderButton(document.getElementById("signInDiv"), {
      theme: "outline",
      size: "large",
    });
  }, []);

  const responseFacebook = (response) => {};

  const logGoogleUser = async () => {
    const { user } = await signInWithGooglePopup();
  };

  return (
    <>
      {
        <All secondScreen={secondScreen}>
          <PageContainer>
            <Photo></Photo>
            <LogInContainer>
              <AuthContainer secondScreen={secondScreen}>
                <AuthContainerDiv secondScreen={secondScreen}>
                  <AuthHeader>Sign Up</AuthHeader>
                  <AuthParagraph>
                    By continuing, you agree are setting up a Reddit account and
                    agree to our <Privacy>User Agreement</Privacy> and{" "}
                    <Privacy>Privacy Policy</Privacy>.
                  </AuthParagraph>
                  <SignInWithGoogle>
                    {/* <div id="signInDiv"></div> */}
                    <button onClick={() => logGoogleUser()}>
                      <FcGoogle size={22} />
                      <span> CONTINUE WITH GOOGLE</span>
                    </button>
                  </SignInWithGoogle>

                  <SignInWithFacebook>
                    <FacebookLogin
                      appId="648981979946188"
                      autoLoad={false}
                      fields="name,email,picture"
                      callback={responseFacebook}
                      cssClass="my-facebook-button-class"
                      icon=<FaFacebookSquare />
                      textButton="CONTINUE WITH FACEBOOK"
                    />
                  </SignInWithFacebook>

                  <OrHeader>
                    <span>OR</span>
                  </OrHeader>

                  <form onSubmit={handleSubmit}>
                    <FormInputPageCom
                      valid={validEmail}
                      initialFocus={initialFocus}
                      showIcon={true}
                      label="EMAIL"
                      type="email"
                      required
                      onChange={handleChange}
                      name="email"
                      value={email}
                      onFocus={() => {
                        setInitialFocus(false);
                      }}
                    />

                    <ErrorParagraph valid={validEmail || initialFocus}>
                      {errMsg}
                    </ErrorParagraph>

                    <ButtonsContainer>
                      {validEmail && (
                        <Button
                          valid={validEmail}
                          page={true}
                          type="button"
                          onClick={() => checkEmail(email)}
                        >
                          Continue
                        </Button>
                      )}
                      {!validEmail && (
                        <Button
                          page={true}
                          disabled
                          valid={validEmail}
                          type="button"
                        >
                          Continue
                        </Button>
                      )}
                    </ButtonsContainer>
                    <br></br>

                    <Forget>
                      Already a redditor?{" "}
                      <button onClick={() => {}}>LOG IN</button>
                    </Forget>
                  </form>
                </AuthContainerDiv>
              </AuthContainer>
            </LogInContainer>
          </PageContainer>
        </All>
      }
    </>
  );
};

export default SignUpPageFirstScreen;
