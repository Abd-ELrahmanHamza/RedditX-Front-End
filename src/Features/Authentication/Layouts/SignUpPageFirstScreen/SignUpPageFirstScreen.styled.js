import styled from "styled-components";

const subColor = "#797b7d";
const blueValid = "#24a0ed";
const transition = "400ms ease all";

const linkColor = "#0079d3";
const facebookBlue = "#237df3";
const errorColor = "#ea0027";

const fontFamily = "'IBM Plex Sans', sans-serif;";

export const All = styled.div`
  display: ${(p) => (p.secondScreen ? "none" : "block")};
`;
export const AuthContainer = styled.div`
  display: flex;
  flex-direction: column;
  max-width: 280px;
  margin: 0px auto 0;
  position: relative;
  display: ${(p) => (p.secondScreen ? "none" : "block")};
  width: 280px;
`;

export const BackSpan = styled.span`
  margin-bottom: 25px;
  position: absolute;
  top: -10%;

  svg {
    width: 25px;
    height: 25px;
    color: ${subColor};
  }
`;

export const AuthContainerDiv = styled.span`
  display: ${(p) => (p.secondScreen ? "none" : "block")};
`;

export const Group = styled.div`
  position: relative;
  margin: 10px 0;
  display: flex;
  justify-content: center;
  align-items: center;
  width: ${(p) => (p.secondScreen ? "350px" : "auto")};

  input[type="password"] {
    letter-spacing: 0.3em;
  }

  span {
    position: absolute;
    right: 5%;
    transition: ${transition};
    color: ${blueValid};
  }
`;

export const PageContainer = styled.div`
  display: flex;
  justify-content: center;
`;

export const Photo = styled.div`
  background-image: url(https://www.redditstatic.com/accountmanager/bbb584033aa89e39bad69436c504c9bd.png);
  width: 156px;
  height: 100vh;
  min-height: 430px;
  background-repeat: no-repeat;
  background-size: cover;
`;

export const LogInContainer = styled.div`
  width: calc(100% - 156px);
  display: flex;
  align-items: center;

  > div {
    margin: 0 0 0 20px;
  }
`;

export const Privacy = styled.span`
  color: ${linkColor};
`;
export const AuthHeader = styled.h2`
  font-family: ${fontFamily};
  font-weight: 500;
  font-size: 20px;
  line-height: 24px;
  max-width: 280px;
  text-transform: none;
`;
export const AuthParagraph = styled.div`
  color: #1a1a1b;
  font-family: Noto Sans, sans-serif;
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
`;

export const ButtonsContainer = styled.div`
  display: flex;
  justify-content: space-between;
`;
export const ErrorParagraph = styled.p`
  color: ${errorColor};
  margin-top: ${(p) => (p.valid ? "-30px" : "0")};
  font-size: 12px;
  transition: all 0.2s ease-in-out;
  padding: 0 0 0 20px;
`;

export const UserNameInput = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const SignInWithGoogle = styled.div`
  display: flex;
  justify-content: center;
  width: 280px;


  margin: 20px 0 5px 0;

  transform: scale(1.19);
  div#singInDiv {
    width:100% !important;
    display: none !important;

   
  }
  iframe {
    min-width:100%;
    border: none;
  }
  
  }
`;

export const SignInWithFacebook = styled.div`
display: flex;
justify-content: center;
width: 280px;


margin: 20px 0 5px 0;



span {
  width: 100%;
 
  button{
    transition: ${transition}
    position: relative;
    width: 100%;
    border-radius: 5px;
    background-color: white;
    height: 47px;
    border-color: ${facebookBlue};
    color: ${facebookBlue};
    cursor: pointer;
    font-size: 14px;
    font-weight: 600;
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    outline: none;
    padding: 0 0 0 20px;

    svg{
      
        width: 40px;
        height: 60px;
        position: absolute;
        left: 1%;

        path{
          color: ${facebookBlue};
        }
        
    }


    &:hover {
      background-color: ${facebookBlue};
      color: white;
      border: none;
      path{
        color: white;
      }

    }
  }
}


}
`;

export const OrHeader = styled.h2`
  width: 100%;
  text-align: center;
  border-bottom: 1px solid #93908e;
  line-height: 0.1em;
  margin: 30px 0;
  line-height: 0 !important;
  color: #93908e;
  font-size: 12px !important;

  span {
    background: #fff;
    padding: 0 10px;
  }
`;
export const Forget = styled.span`
  font-size: 12px;

  button {
    padding: 0;
    font-weight: 700;
    border: none;
    background-color: transparent;
    color: #0079d3;
    font-family: ${fontFamily};
  }
`;
export const FirstDiv = styled.div`
  min-height: calc(14vh - 20px);
  margin: 20px 0 0 20px;
`;
export const MidDiv = styled.div`
  display: flex;
  justify-content: space-between;
  min-height: calc(69vh - 20px);
  padding: 20px 20px 0 20px;
`;
export const LastDiv = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  min-height: calc(14vh - 20px);
  padding: 0 20px 0 20px;
  background-color: #f6f7f8;
  position: fixed;
  bottom: 0;
  left: 0;
  width: 100%;
`;
export const BackButton = styled.button`
  border: none;
  background-color: transparent;
`;
export const RandomContainer = styled.div`
  p {
    color: ${linkColor};
    cursor: pointer;

    &:hover {
        color: ${blueValid};
    }
  }
`;
export const Reload = styled.span`
  margin-left: 10px;
  font-size: 18px;
  cursor: pointer;
  color: ${blueValid};
`;
