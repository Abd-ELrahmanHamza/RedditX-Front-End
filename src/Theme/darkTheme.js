/**
 * dark theme colors
 *
 * @type {{id:string}}
 */
const darkTheme = {
  // The id of the theme
  id: "dark",
  // The canvas color is the background color of the website
  canvas: "#030303",
  // The colors of the text on the website
  color: {
    // The primary color of the text on the website
    primary: "#d7dadc",
    // The color of the muted text on the website
    muted: "#7c7c7c",
    // The color of the links on the website
    link: "#0079d3",
  },
  // The colors of the backgrounds in the website
  background: {
    primary: "#1a1a1b",
  },
  // The colors of the lines in the website
  lineColor: {
    primary: "#343536",
  },
};

export default darkTheme;
