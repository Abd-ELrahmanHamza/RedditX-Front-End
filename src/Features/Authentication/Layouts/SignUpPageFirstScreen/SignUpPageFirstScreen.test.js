import { shallow } from "enzyme";

import SignUpPageFirstScreen from "./SignUpPageFirstScreen";
import { BrowserRouter } from "react-router-dom";

import { ThemeProvider } from "styled-components";

import darkTheme from "Theme/darkTheme";
import { screen, render, fireEvent, waitFor } from "@testing-library/react";



const validEmail = false;
const secondScreen = false;
const initialFocus = false;
const initialFocus2 = false;
const setValidEmail = jest.fn();
const setFormFields = jest.fn();
const setSecondScreen = jest.fn();
const setInitialFocus = jest.fn();
const setInitialFocus2 = jest.fn();
const setErrMsg = jest.fn();
const setSug1 = jest.fn();
const setSug2 = jest.fn();
const setSug3 = jest.fn();
const setSug4 = jest.fn();
const setSug5 = jest.fn();


const defaultFormFields = {
  userName: "",
  password: "",
  email: "",
};

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe("SignUpPageFirstScreen Layout", () => {
  it("this is a test for SignUpPageFirstScreen Layout", () => {
    expect(
      shallow(
        <BrowserRouter>
          <SignUpPageFirstScreen formFields={defaultFormFields} />
        </BrowserRouter>
      )
    ).toMatchSnapshot();
  });



  test("email-error shouldn't be found", () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
          <SignUpPageFirstScreen
            validEmail={true}
            secondScreen={secondScreen}
            initialFocus={true}
            initialFocus2={initialFocus2}
            setValidEmail={setValidEmail}
            setFormFields={setFormFields}
            setSecondScreen={setSecondScreen}
            setInitialFocus={setInitialFocus}
            setInitialFocus2={setInitialFocus2}
            setErrMsg={setErrMsg}
            formFields={defaultFormFields}
            setSug1={setSug1}
            setSug2={setSug2}
            setSug3={setSug3}
            setSug4={setSug4}
            setSug5={setSug5}
          />
        </ThemeProvider>
      </BrowserRouter>
    );
    expect(screen.queryByTestId("email-error")).not.toBeVisible();
  });

  test("email-error should be found", async () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
        <SignUpPageFirstScreen
            validEmail={validEmail}
            secondScreen={secondScreen}
            initialFocus={initialFocus}
            initialFocus2={initialFocus2}
            setValidEmail={setValidEmail}
            setFormFields={setFormFields}
            setSecondScreen={setSecondScreen}
            setInitialFocus={setInitialFocus}
            setInitialFocus2={setInitialFocus2}
            setErrMsg={setErrMsg}
            formFields={defaultFormFields}
            setSug1={setSug1}
            setSug2={setSug2}
            setSug3={setSug3}
            setSug4={setSug4}
            setSug5={setSug5}
          />
        </ThemeProvider>
      </BrowserRouter>
    );
    const emailInput = screen.getByTestId("email");

    fireEvent.change(emailInput, {
      target: { value: "M" },
    });

    emailInput.focus();

    const emailError = screen.getByTestId("email-error");

    await waitFor(() => {
      expect(emailError).toBeVisible();
    });
  });
  test("email-error should be not found", async () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
        <SignUpPageFirstScreen
            validEmail={validEmail}
            secondScreen={secondScreen}
            initialFocus={true}
            initialFocus2={initialFocus2}
            setValidEmail={setValidEmail}
            setFormFields={setFormFields}
            setSecondScreen={setSecondScreen}
            setInitialFocus={setInitialFocus}
            setInitialFocus2={setInitialFocus2}
            setErrMsg={setErrMsg}
            formFields={defaultFormFields}
            setSug1={setSug1}
            setSug2={setSug2}
            setSug3={setSug3}
            setSug4={setSug4}
            setSug5={setSug5}
          />
        </ThemeProvider>
      </BrowserRouter>
    );
    const emailInput = screen.getByTestId("email");

    fireEvent.change(emailInput, {
      target: { value: "Mgfdgs@gmail.com" },
    });

    const emailError = screen.getByTestId("email-error");

    await waitFor(() => {
      expect(emailError).not.toBeVisible();
    });
  });

  test("email-input should be found", () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
        <SignUpPageFirstScreen
            validEmail={validEmail}
            secondScreen={secondScreen}
            initialFocus={initialFocus}
            initialFocus2={initialFocus2}
            setValidEmail={setValidEmail}
            setFormFields={setFormFields}
            setSecondScreen={setSecondScreen}
            setInitialFocus={setInitialFocus}
            setInitialFocus2={setInitialFocus2}
            setErrMsg={setErrMsg}
            formFields={defaultFormFields}
            setSug1={setSug1}
            setSug2={setSug2}
            setSug3={setSug3}
            setSug4={setSug4}
            setSug5={setSug5}
          />
        </ThemeProvider>
      </BrowserRouter>
    );
    const emailInput = screen.getByTestId("email");
    expect(emailInput).toBeInTheDocument();
  });
});
