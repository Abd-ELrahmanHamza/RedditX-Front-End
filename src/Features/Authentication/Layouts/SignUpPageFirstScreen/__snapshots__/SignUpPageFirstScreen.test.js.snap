// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpPageFirstScreen Layout this is a test for SignUpPageFirstScreen Layout 1`] = `ShallowWrapper {}`;
